GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

EddieChoCho opened the pull request:
Fixes: #3444

* Add `DistributedLock` interface which is implemented by `RedisLock` and `JdbcLock`.
* Modify `LockRegistry`, `ExpirableLockRegistry`, `RenewableLockRegistry` interfaces.
* Modify implementation of `DefaultLockRepository, `JdbcLockRegistry`, `RedisLockRegistry`
* Modify ddl of `INT_LOCK` table.
* Maintain test cases and documents.

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

EddieChoCho commented on spring-integration-core/src/main/java/org/springframework/integration/support/locks/CustomTtlLock.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @since 6.3
+ */
+public interface CustomTtlLock extends Lock {
```
In this PR, the `CustomTtlLock` interface is only implemented by `RedisLock` and `JdbckLock`.
It can be renamed to `DistributedLock` in the future if all the implementations of Spring distributed lock have implemented the interface.

artembilan reviewed (changes requested):
Thanks for looking into this!

I still think it has to be called a `DistributedLock` from the beginning: we may come up eventually with other API, not only TTL.

I think the `LockRegistry` must be modified for a generic arg: `LockRegistry<L extends Lock>`.
And no need in the extra `CustomTtlLockRegistry` abstraction.

Can you elaborate, please, why do we need a result from `delete()`?

Since this is a huge braking change (as it is expected according to the discussion in the issue), we are going to look into this in details in the next `6.4` version, which we are going to start this June.

EddieChoCho commented:
> Can you elaborate, please, why do we need a result from `delete()`?

If the result(row count) is 0, which means that the lock is not owned by current process.

For example:
ProcessA and processB are in the same region. 
ProcessA acquired the distributed lock. //{REGION: r, LOCK_KEY: lock, CLIENT_ID: processA}
But the work takes too long and the ownership of the lock has been expired.
Before processA release the lock, processB acquired the same distributed lock.  //{REGION: r, LOCK_KEY: lock, CLIENT_ID: processB}
When processA try to release the distributed lock. //DELETE FROM INT_LOCK WHERE REGION='r' AND LOCK_KEY='lock' AND CLIENT_ID='processA'
The result from `delete()` would be 0.

In this case, an IllegalStateException should be thrown to info the process that the integrity of data protected by this lock may have been compromised.

artembilan commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java:
```diff
@@ -305,12 +351,21 @@ public void unlock() {
 						// try again
 					}
+					catch (IllegalStateException e) {
+						throw new IllegalStateException("Lock was released in the store due to expiration. " +
```
OK. I see why you use `boolean` for `delete()` now.
I wonder if we can use a `ConcurrentModificationException` instead.
Or something from the `org.springframework.dao` package, like `ConcurrencyFailureException`

artembilan replied:
I see. Well, that was just a suggestion if that is makes sense to rely on some specific exception instead of this common.
Might be the fact that `ConcurrentModificationException` does not fit in our locks scenario.
However that `ConcurrencyFailureException` might be OK.

EddieChoCho replied:
```
catch (TransientDataAccessException | TransactionTimedOutException | TransactionSystemException e) {
      // try again
}
```
In the current implementation, `TransientDataAccessException`, `TransactionTimedOutException`, and `TransactionSystemException` will be caught, and then the `delete()` method will be invoked again.

Since `ConcurrencyFailureException` extends `TransientDataAccessException`, I think we should not throw it for this scenario.

I will change it to throw `ConcurrentModificationException` instead if you think `ConcurrentModificationException` fits better.

artembilan replied:
I see. Thanks. Missed that.

So, yeah, this `ConcurrentModificationException` fits in my brain because its name means exactly what happens here for us.
But I'm not sure if this is a correct scenario where we can use it from Java perspective.
That's why I'm reaching you for other opinion, arguments and so on.
Just want to be sure that we are on the same page with a decision to move on.

EddieChoCho replied:
I see, 
I agree that changing from `IllegalStateException` to `ConcurrentModificationException` can improve the readability.
But I'm also not sure if this is a correct scenario where we can use it from Java perspective.

How about we create our own exception -`LockOwnershipExpiredException` for Spring distributed locks?
It can be used by both `RedisLock`, and `JdbcLock`.

artembilan commented:
This is good stuff. 
Thank you very much for looking into that!

As I expected this is really huge breaking change in the API, so I'm afraid we cannot accept it right now according to our migration policies.
We have to wait until the next major version `7.0` next year.
Of course we might be able to start it in the end of this year, when Spring Framework `6.2` release train is out this November.

Please, keep in touch meanwhile.

Thank you for understanding!

Pavankn18 commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/DefaultLockRepository.java:
```diff
@@ -389,23 +375,23 @@ public void close() {
-	public void delete(String lock) {
-		this.defaultTransactionTemplate.executeWithoutResult(
-				transactionStatus -> this.template.update(this.deleteQuery, this.region, lock, this.id));
+	public boolean delete(String lock) {
```
This also presents an opportunity to check for `template.update() > 1` to throw exception, this will help prevent incorrect constraint on Table. 

This is possible only if table is created without `constraint INT_LOCK_PK primary key (LOCK_KEY, REGION)`.

EddieChoCho replied:
This is actually outside the scope of the PR. (my bad, I should not include this change). 
The method has been maintained in PR #9292. 
I think business logic should not address cases of misconfiguration or incorrect settings. 
If you'd like to discuss this method further, it would be great to create a new issue.

Pavankn18 commented:
Any new on this? Spring 7 will be out soon right? Will these changes be included 7.0?

artembilan commented:
Thank you both for keeping an eye on this! So, in November we are releasing `6.4`. After that we are good to start `7.0`. But that is going to be released only next November 2025. So, not soon enough, but we still don’t have a choice with such a breaking change.

EddieChoCho commented:
I had rebased the branch, here is the summary of the changes:
* `RedisLock` and `JdbcLock` now implement `DistributedLock` interface to support the feature of customized time-to-live for the lock status data.
* `RenewableLockRegistry` now provides new `renewLock(Object lockKey, long customTtl, TimeUnit customTtlTimeUnit)` method, allowing user to renew the lock with a custom time-to-live value.
* The APIs of `LockRepository` and `DefaultLockRepository` are also modified to support the feature.
* The `CREATED_DATE` column in the `INT_LOCK` table has been replaced with `EXPIRED_AFTER` to support the feature.

EddieChoCho committed:
GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

Fixes: #3444

* Add `DistributedLock` interface which is implemented by `RedisLock` and `JdbcLock`.
* Modify `LockRegistry`, `ExpirableLockRegistry`, `RenewableLockRegistry` interfaces.
* Modify implementation of `DefaultLockRepository, `JdbcLockRegistry`, `RedisLockRegistry`
* Modify ddl of `INT_LOCK` table.
* Maintain test cases and documents.

Signed-off-by: Eddie Cho <chochoforwork@gmail.com>

cppwfs reviewed (changes requested):
Thank you for the contribution.  

Only had a couple of small requests.

artembilan commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java:
```diff
@@ -85,12 +86,25 @@ protected boolean removeEldestEntry(Entry<String, JdbcLock> eldest) {
+		this.ttl = DEFAULT_TTL;
+	}
+
+	public JdbcLockRegistry(LockRepository client, long expireAfter) {
```
Let's see if we can rely on `Duration` instead!

EddieChoCho replied:
Let's also update the constructor of `RedisLockRegistry` from
`public RedisLockRegistry(RedisConnectionFactory connectionFactory, String registryKey, long expireAfter)`
to
`public RedisLockRegistry(RedisConnectionFactory connectionFactory, String registryKey, Duration expireAfter)`
to align with the API design.

Do you agree?

artembilan replied:
Sure! We can do that, but as a new constructor and deprecation of existing one.
Remember, we develop here a library, therefore compatibility in API should be honored as much as possible.
Since we can still keep an old constructor here with simple adaptation, no reason to introduce another breaking change.

artembilan commented on spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/schema-db2.sql:
```diff
@@ -30,7 +30,7 @@ CREATE TABLE INT_LOCK  (
 	REGION VARCHAR(100) NOT NULL,
 	CLIENT_ID CHAR(36),
-	CREATED_DATE TIMESTAMP NOT NULL,
+    EXPIRED_AFTER TIMESTAMP NOT NULL,
```
Something is off with indents.
Has to be tab.

artembilan reviewed (changes requested):
This is really great!
Please, take a look into `distributed-locks.adoc` to mention this our new `DistributedLock` abstraction and fix code snippet for a new style: `LockRegistry<Lock>` and so.

Thank you!

EddieChoCho commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/DefaultLockRepository.java:
```diff
@@ -189,14 +182,6 @@ public void setPrefix(String prefix) {
-	 * Specify the time (in milliseconds) to expire deadlocks.
-	 * @param timeToLive the time to expire deadlocks.
-	 */
-	public void setTimeToLive(int timeToLive) {
```
I moved the default TTL constant from `DefaultLockRepository` to `JdbcLockRegistry`.
The TTL value can now be set through the constructor of `JdbcLockRegistry`, which aligns with the design of `RedisLockRegistry`.
As a result, I removed the `setTimeToLive()` method from `DefaultLockRepository`.

However, should I also keep a default TTL in `DefaultLockRepository` to support developers who use `DefaultLockRepository` directly without using `JdbcLockRegistry`?

EddieChoCho committed:
GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

Fixes: #3444

* Modify `LockRegistry`, `DistributedLock` interfaces.
* Modify implementation of `DefaultLockRepository, `JdbcLockRegistry`, `RedisLockRegistry`
* Modify ddl of `INT_LOCK` table.
* Maintain test cases and documents.

Signed-off-by: Eddie Cho <chochoforwork@gmail.com>

EddieChoCho commented:
I just pushed another commit, please review

artembilan reviewed (changes requested):
I don't see too many problem now.
Thank you for the great update!
After the next round of changes according to our review, we are going to look into merging this.

EddieChoCho committed:
GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

Fixes: #3444

* Modify DistributedLock` interfaces.
* Modify implementation of `JdbcLockRegistry`, `RedisLockRegistry`.
* Modify JavaDoc.

Signed-off-by: Eddie Cho <chochoforwork@gmail.com>

EddieChoCho committed:
GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

Fixes: #3444

* Maintain document

Signed-off-by: Eddie Cho <chochoforwork@gmail.com>

EddieChoCho committed:
GH-3444: Add Custom TTL support for RedisLock, and JdbcLock

Fixes: #3444

* Add @Deprecated

Signed-off-by: Eddie Cho <chochoforwork@gmail.com>

cppwfs commented:
LGTM!    Thank you for your contribution!!!!

cppwfs commented:
Squashed, Rebased, Merged

Thank you for the hard work on this contribution!
